src/Dumplie/Domain/SharedKernel/Money/Price.php

```diff
@@ -0,0 +1,119 @@
+<?php
+
+declare (strict_types = 1);
+
+namespace Dumplie\Domain\SharedKernel\Money;
+
+use Dumplie\Domain\SharedKernel\Exception\DifferentPricePrecisionException;
+use Dumplie\Domain\SharedKernel\Exception\InvalidArgumentException;
+use Dumplie\Domain\SharedKernel\Exception\InvalidCurrencyException;
+
+final class Price
+{
+    const DEFAULT_PRECISION = 100;
+
+    /**
+     * @var int
+     */
+    private $amount;
+
+    /**
+     * @var string
+     */
+    private $currency;
+
+    /**
+     * @var int
+     */
+    private $precision;
+
+    /**
+     * @param int $amount Amount, expressed in the smallest units of $currency (eg cents)
+     * @param string $currency
+     * @param int $precision used to calculate float value, 10000 cents / 100 = 100.00
+     * @throws InvalidArgumentException
+     */
+    public function __construct(int $amount, string $currency, int $precision = self::DEFAULT_PRECISION)
+    {
+        if ($amount < 0 || $precision < 0) {
+            throw new InvalidArgumentException;
+        }
+
+        $this->amount = $amount;
+        $this->currency = $currency;
+        $this->precision = $precision;
+    }
+
+    /**
+     * @param $method
+     * @param $arguments
+     * @return Price
+     */
+    public static function __callStatic($method, $arguments) : Price
+    {
+        return new Price($arguments[0], $method, isset($arguments[1]) ? $arguments[1] : self::DEFAULT_PRECISION);
+    }
+
+    /**
+     * @param int $value
+     * @param string $currency
+     * @return Price
+     */
+    public static function fromInt(int $value, string $currency) : Price
+    {
+        return new Price($value * self::DEFAULT_PRECISION, $currency);
+    }
+
+    /**
+     * @return float
+     */
+    public function floatValue() : float
+    {
+        return $this->amount / $this->precision;
+    }
+
+    /**
+     * @return string
+     */
+    public function currency() : string
+    {
+        return mb_strtoupper($this->currency);
```

Comment:
hasSameCurrency will always compare uppercase currency code, even if you pass it as a lowercase.